lib/assert.js

```diff
@@ -140,95 +139,154 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
-  // properties (`global`, `multiline`, `lastIndex`, `ignoreCase`).
-  } else if (util.isRegExp(actual) && util.isRegExp(expected)) {
-    return actual.source === expected.source &&
-           actual.global === expected.global &&
-           actual.multiline === expected.multiline &&
-           actual.lastIndex === expected.lastIndex &&
-           actual.ignoreCase === expected.ignoreCase;
-
-  // If both values are primitives, equivalence is determined by
-  // == or, if checking for strict equivalence, ===.
-  } else if ((actual === null || typeof actual !== 'object') &&
-             (expected === null || typeof expected !== 'object')) {
+  // For primitives / functions
+  // (determined by typeof value !== 'object'),
+  // or null, equivalence is determined by === or ==.
+  if (isNullOrNonObj(actual) && isNullOrNonObj(expected)) {
     return strict ? actual === expected : actual == expected;
+  }
 
-  // If both values are instances of typed arrays, wrap their underlying
-  // ArrayBuffers in a Buffer to increase performance.
-  // This optimization requires the arrays to have the same type as checked by
-  // Object.prototype.toString (pToString). Never perform binary
-  // comparisons for Float*Arrays, though, since +0 === -0 is true despite the
-  // two values' bit patterns not being identical.
-  } else if (ArrayBuffer.isView(actual) && ArrayBuffer.isView(expected) &&
-             pToString(actual) === pToString(expected) &&
-             !(actual instanceof Float32Array ||
-               actual instanceof Float64Array)) {
-    return compare(Buffer.from(actual.buffer,
-                               actual.byteOffset,
-                               actual.byteLength),
-                   Buffer.from(expected.buffer,
-                               expected.byteOffset,
-                               expected.byteLength)) === 0;
-
-  // For all other Object pairs, including Array objects, equivalence is
-  // determined by having the same number of owned properties (as verified
-  // with Object.prototype.hasOwnProperty.call), the same set of keys
-  // (although not necessarily the same order), equivalent values for every
-  // corresponding key, and an identical 'prototype' property. Note: this
-  // accounts for both named and indexed properties on Arrays.
-  } else {
-    memos = memos || {actual: [], expected: []};
+  // If they bypass the previous check, then at least
+  // one of them must be an non-null object.
+  // If the other one is null or undefined, they must not be equal.
+  if (actual === null || actual === undefined ||
+      expected === null || expected === undefined)
+    return false;
 
-    const actualIndex = memos.actual.indexOf(actual);
-    if (actualIndex !== -1) {
-      if (actualIndex === memos.expected.indexOf(expected)) {
-        return true;
-      }
+  // Notes: Type tags are [[Class]] properties returned by
+  // obj->class_name() in C++ or Object.prototype.toString.call(obj) in JS
```

Comment:
I mentioned `obj->class_name()` because that looks like it's what `util.isDate()` (`Value::IsDate`) and alike come down to in a macro. Also it's why `process` and a lot of other objects created in the C++ land has their own tags, although after a quick search looks like they actually use `FunctionTemplate::SetClassName()` to set this, and never use the getter, so probably we can just mention `FunctionTemplate::SetClassName()` would alter this.

`[[Class]]` is described as "historical" in another PR, just as it is in the spec note.